### ISSUE TITLE
Implement ordinal regression structures

### ISSUE THREAD

**Issue by StaffanBetner** [glmmTMB/glmmTMB#514]
It would be very useful if glmmTMB supported different types of ordinal regression, e.g. the proportional odds model (cumulative link logit model).

**Comment by StaffanBetner** [comment]
The ordinal package is a bit of a mess, unfortunately. The old implementation (cl(m)m2) only support one random intercept, and the new one doesn't have any predict method (which makes it incompatible with DHARMa) or scale/nominal effects. It also does not support REML estimation.

**Comment by bbolker** [comment]
Fair enough.  On the other hand:

- the ordinal package is quite mature and well-tested
- Implementing a predict method for clmm in the ordinal package would probably be easier than writing an implementation of ordinal models from scratch
- REML would probably be easier to implement in glmmTMB, but my personal opinion is that its value is a little bit overrated ...

**Comment by willtudorevans** [comment]
Just to second the request for `glmmTMB` to add ordinal regression to its supported families. I'm finding the `ordinal` package can't work with large datasets in the way `glmmTMB` can.

**Comment by mmrabe** [comment]
Moreover, the `clmm` function currently doesn't support scale effects. The older `clmm2` did but only supports one random factor. So ordinal regression models with more than one random factor *and* scale effects cannot be fitted using the `ordinal` package. So I'd also like to second this request. :-)

**Comment by mmrabe** [comment]
I'd also like to second this request. Fitting ordinal regression models with scale effects and more than one random factor would be a big plus. I'm not aware of any R package that would be capable of this within a frequentist modelling framework.

**Comment by strengejacke** [comment]
Can I revive this old issue? Given how fast and easy it seemed to implement ordered beta regression, maybe the underlying TMB package has has developed so well that it is now easier to implement ordinal or multinomial regression?

**Comment by phalaropus** [comment]
Ordinal regression structures in glmmTMB would welcomed by ecologists and others.

**Comment by bbolker** [comment]
Chiming back in:

- I agree that having ordinal models in `glmmTMB` would be useful, although:
    - `clmm` does already support multiple REs
    - I'm not entirely clear what "scale effects" mean in this context?
    - mildly surprising that `ordinal` does poorly on large data sets
- ordered-beta was easier to implement than ordinal would be &mdash; all it required was implementing the log-likelihood function, and allowing two shape parameters in the model. For ordinal we need to deal with issues like prediction, simulation, etc etc ...
- I would welcome a pull request!

**Comment by mmrabe** [comment]
Thanks for taking another look, @bbolker! I understand that this is quite demanding but I just wanted to make sure to explain what I meant by "scale effects":

Those are basically effects on the residual variance (or rather on the SD), e.g. to allow the residual variance to differ between experimental conditions. `ordinal::clm` (fixed effects only) and `ordinal::clmm2` (mixed effects but only a single random factor) implement this but the most advanced `ordinal::clmm` does not.

I had a particular model in mind with an ordinal response variable (confidence), two crossed random factors, and residual variance that systematically differs between two levels of a fixed factor. That is neither suitable for `ordinal::clmm` because that function has no "scale effects", nor for `ordinal::clmm2` or `ordinal::clm` because the model has two random factors.

**Comment by qdread** [comment]
I would like to second this. We are specifically looking for a frequentist implementation in R of the following model:

- Cumulative logit model
- Proportional odds assumption is violated (i.e. we need category-specific effects, otherwise known as nominal effects)
- Two nested random effects with 0 covariance
- No scale effects as @mmrabe mentioned above

This is currently not possible with `ordinal::clmm2`, which allows the category-specific effects but not nested random effect design, nor with `ordinal::clmm`, which allows the nested random effect design but not the category-specific effects. I have coded the model in Stan, and there is also support for it in brms. However there is no frequentist implementation in R that I can find. I have the same problem as @mmrabe where `clmm2` has support for esoteric versions of ordinal multinomial models (scale effects and nominal effects) but `clmm` has support for more complex random effect structures. 

I would be interested in trying to learn how to implement my own family in glmmTMB to do this, but I do not have the experience in C++ coding to do that. Any input would be greatly appreciated.

**Comment by bbolker** [comment]
I haven't thought about this in a little while, but I think that this can be nearly entirely implemented on the R side, by constructing specialized $\mathbf X$ and $\mathbf Z$ matrices (and maybe creating pseudo-observations) ... in which case very little C++ coding (maybe none?)  Internally, I think the model would look like a binomial where the logit link would take care of the baseline proportional odds assumption. If I'm right, the hard part would be dealing with all of the implicit general assumptions we're making that would be broken by this new family ...

What does a minimal implementation look like in Stan?

**Comment by qdread** [comment]
Thanks for thinking about it! Here is an implementation of the model discussed in Chapter 14 of [Stroup et al.'s GLMM book](https://www.routledge.com/Generalized-Linear-Mixed-Models-Modern-Concepts-Methods-and-Applications/Stroup-Ptukhina-Garai/p/book/9781498755566) in Stan. I have provided Stan code for the proportional odds model and the model where proportional odds assumption is relaxed (i.e. there are category-specific, AKA nominal, effects). Both models have nested random effects.

https://github.com/qdread/stan_multinomial_poinsettia

The repo has the two Stan scripts, example dataset, and R script to fit the Stan models. The example dataset is supposed to represent poinsettias of 3 different varieties grown by 12 different growers that are rated on an ascending scale C, B, A. Variety is the treatment and grower is the block.

**Comment by bbolker** [comment]
Embarrassingly, I don't have a copy of Stroup's book (I need to get one).  Your Stan code is a good start but for glmmTMB we need to figure out how to implement everything so that it (1) generalizes nicely to any number of levels; (2) works mostly by linear algebra/matrix computation rather than indexing parameter vectors ... I may be able to start from your code and get to something standalone based on matrix computation, then think about how to incorporate it into glmmTMB without breaking anything ...

**Comment by qdread** [comment]
Thanks for taking a look! I will try to work on the Stan implementation to turn the loops + indexing into a proper matrix computation, and update you if/when I have something working.

**Comment by qdread** [comment]
Hi again, I have updated the [Stan code in the repo](https://github.com/qdread/stan_multinomial_poinsettia). I've turned the fixed and random effects into design matrices. Now the linear predictors are made by summing up the products of matrix multiplications. So that part is now based on matrix computation and is generalizable to however many levels. I also rewrote the inverse link part so that it is generalizable to >3 levels, though there may be a better way to do that part. Let me know if there is any other way I can help.

**Comment by bbolker** [comment]
FWIW, from the [ordinal package vignette](https://cran.r-project.org/web/packages/ordinal/vignettes/clm_article.pdf):

> The restriction of the threshold parameters {θj } being non-decreasing is dealt with by defining ℓ(θ, β; y) = ∞ when {θj } are not in a non-decreasing sequence. If the algorithm attempts evaluation at such illegal values step-halving effectively brings the algorithm back on track. Other implementations of CLMs re-parameterize {θj } such that the non-decreasing nature of {θj } is enforced by the parameterization, for example, `MASS::polr` (package version 7.3.49) optimize[s] the likelihood using 

$$
\tilde θ_1 = θ1, \tilde θ_2 = \exp(θ_2 − θ_1), . . . , \tilde θ_{J−1} = \exp(θ_{J−2} − θ_{J−1})
$$

> This is deliberately not used in ordinal because the log-likelihood function is generally closer to quadratic in the original parameterization in our experience which facilitates faster convergence.

**Comment by bbolker** [comment]
Some thoughts about implementation.

* possibly the most annoying part is the need to keep track of $\theta$ and $\beta$ separately, since they act differently/have different roles. We could either:
   * store the $\theta$ values as (say) the first segment of the $\beta$ vector, and then make sure the proper subset is selected whenever we want to do computations like $X \beta$
   * add another (generally unused) vector to the list of parameter vectors
   *  store the $\theta$ vectors in the "extra family specific parameters" vector $\psi$
I'm leaning toward the last option.
* Suppose we do assume that $\psi$ (family parameters) are the intercept parameters (not sure whether we have to explicitly suppress an intercept in the $\beta$/$X$ parameterization? Then (pseudocode) something like this:

```
psicum = cumsum(psi)  ## or c(cumsum(psi), Inf) ?
prob[i] <- {
   if (response[i] == maxresponse) {
       1 - linkinv(psicum[maxresponse]
   } else {
       diff(linkinv(psicum(response[c(i, i-1)]))
  }
}
log(prob[i])
```
* next step is probably to implement something like this in R (or RTMB) and get the results matching with `ordinal::clm` or `MASS::polr`
* for numerical stability we would want to replace the `diff()` operator with something that directly computed the log probability from the differences on the link scale ... (`diff(response)*dmu_deta` for extreme values?)
* I haven't thought too much about what we'd have to do to get scale or nominal effects.

**Comment by DrJerryTAO** [comment]
I emailed the author of `ordinal` months ago but received no response. Its github does not seem to address issues. I have reprogrammed 23 hidden functions beneath `ordinal::clm()` locally for my special application (which does not include random effects). Let me know if one needs help understand the undocumented functions. Comments to @bbolker 's thoughts:

- In `ordinal`, θ (thresholds and nominal effects, referred to as `clm()$alpha`), β (location coefficients, referred to as `clm()$beta`), and scale coefficients (referred to as `clm()$zeta`) are kept in one vector, as shown by in `rho$par` for all parameter estimates in hidden functions `clm.newRho()` and `get_clmRho.default()`. 
- Then a matrix of 0/1 is used to generate design matrix for each response level at dichotomization (`B1, B2` with special offset `o1, o2` to give roughly `-Inf` and `+Inf` (+/-`1e5` actually) for `y == 1` and `y == maxy`, respectively, in `clm.newRho()`).
- Supplied to the optimizer includes a matrix multiplication to get linear predictor for each cumulative probability. See `(drop(B1 %*% par[1:n.psi]) + o1)/sigma` in `clm.nll()` where `sigma` contains the scale effects in `clm(scale = ~)` (which should correspond to `disp = ~` in glmmTMB). 
- Thresholds are effectively constrained nominal effects (i.e., coefficients vary between response levels), so thresholds and nominal effects should be handled together. `ordinal` also allows constraint matrix to be applied to thresholds and nominal effects in `threshold = c("flexible", "symmetric", "symmetric2", "equidistant")`. Once a predictor is included in the `nominal = ~` subformula, its location coefficient will be aliased (if the predictor is also included in the main formula, `summary()$coefficients` show NA) because both cannot be identified together. `ordinal` also runs a test function to assert if threshold + nominal are increasing with response levels across all observations and otherwise gives a warning. Does the situation favor the threshold estimates to be stored in an "extra family specific parameters" vector? 
- To enforce thresholds (not nominal effects) to increase with response levels, `ordinal` starts with default threshold estimates that make all response categories equally probably by `qlogis((1:ntheta)/(ntheta + 1))` in `startthreshold()`. All other parameters (nominal coefficients, location coefficients, scale coefficients) start from 0. The optimizer such as `clm.nll()` has no constraints except `if (all(is.finite(rho$fitted)) && all(rho$fitted > 0)) -sum(rho$wts * log(rho$fitted)) else Inf` to give the loglikelihood function, where `rho$fitted > 0` means that fitted probabilities of observed categories must be positive, equivalently dropping cases where thresholds do not monotonically increase with the response level. 
- Whether we have to explicitly suppress an intercept in the βX parameterization? `ordinal` does so. For example, in `clm.newRho()`, the line `B1 <- cbind(B1, X[keep, -1, drop = FALSE])` drops the intercept (the first column) in the location design matrix (`X`). This 0 intercept but free thresholds seem the most common and well understood choice to ensure identifiability. The developer, however, can add routines when all thresholds are known and supplied by the user (e.g. income brackets), returned estimates are the scale parameter and the intercept that can be inversely inferred from estimated thresholds (only two parameters needed).

**Comment by strengejacke** [comment]
I recently came across some papers that use ordinal regression models in combination with latent profile analysis (LPA)/growth mixture modelling (GMM). First, different groups of trajectories of the outcome are identified using LPA/GMM, and then the group variable is used as dependent variable in a next step using ordinal regression.

What I want to say is that it seems like ordinal regressions are probably quite useful for modern statistics, and a modern modelling package in R that already supports ordinal regression is not available, so it would be great to have it in glmmTMB :-)

I wish I would understand more about the underlying stats/maths so I could possibly help, but... that's not the case, unfortunately.

**Comment by bbolker** [comment]
The hard part here is not (IMO) the underlying stats/maths, but the challenges of incorporating it into `glmmTMB`'s structure without making things horribly complex.  I started working on RTMB implementations [here](https://github.com/bbolker/bbmisc/blob/master/Rmisc/ordinal_RTMB.R), just to see that I understood how everything worked.  Some of the big problems/deviations from the current `glmmTMB` structure are in dealing with the intercept parameters, which have to be handled differently from ordinary parameters (because they have an ordered structure and are incorporated in the linear predictor in a different way ...)

**Comment by jmgirard** [comment]
Following up on this thread: I've implemented a working version of an `ordinal` family (proportional-odds cumulative link) for glmmTMB, following the design @bbolker sketched in the [Oct 2024 comment](https://github.com/glmmTMB/glmmTMB/issues/514#issuecomment-2430571238) above (thresholds in `psi`) and the RTMB prototype in bbolker/bbmisc.

Branch: https://github.com/jmgirard/glmmTMB/tree/ordinal-family ([diff vs. master](https://github.com/glmmTMB/glmmTMB/compare/master...jmgirard:glmmTMB:ordinal-family), single commit, +374/−22)

**Usage:**

```r
fit <- glmmTMB(rating ~ temp + contact + (1 | judge),
               data = wine, family = ordinal())  # logit/probit/cloglog
family_params(fit)             # thresholds, named "1|2", "2|3", ...
predict(fit, type = "probs")   # n x K probability matrix (se.fit supported)
```

**Validation** (all three links):

- `ordinal::clm` / `MASS::polr` fixed-effects (housing data): logLik, coefficients, and thresholds agree to ~1e-5 or better
- `ordinal::clmm` (wine data, `(1|judge)`): logLik, coefs, thresholds, RE sd agree to ~1e-5
- predicted probabilities match `clm` to ~1e-5; simulate-and-recover on n=2000 with random intercepts matches `clmm` to 4 decimals
- 27 new testthat assertions pass; no regressions in the existing suite (basics/families/predict/simulate/methods/disp)

**Design choices, per the Oct 2024 comment — happy to change any of these:**

1. Thresholds live in `psi` (option 3 from that comment), length K−1 (data-dependent, so `find_psi()` is bypassed for this family).
2. Monotonicity is enforced by parameterization: `psi = c(theta[1], log(diff(theta)))`, MASS::polr-style. I know the `ordinal` package deliberately avoids this for convergence reasons; starting values are equiprobable-category thresholds (as in `clm`) and I haven't seen convergence problems in testing, but this is easy to swap for the NaN-on-violation approach if you prefer.
3. The fixed-effect intercept is fixed to 0 via an internal `map` entry (absorbed into the thresholds) rather than dropping the X column — this keeps the model matrices and predict machinery untouched. Cosmetic cost: summary shows an `(Intercept) 0 NA` row.
4. `mu` (and hence `type="response"`, `fitted()`, response residuals) is the expected category index E[Y]; per-category probabilities are REPORTed and exposed via a new `predict(type = "probs")` (ADREPORTed when `se.fit = TRUE`).
5. The response must be an ordered factor (or 1-based integer codes); level labels are stored in `modelInfo$ord_levels`; `simulate()` returns ordered factors on the original levels.
6. Dunn-Smyth residuals are implemented via the discrete PIT of the CLM CDF, computed R-side from the thresholds and latent predictor.

No changes to the TMB data structures (so no `up2date()` burden) and no new family-specific arguments in the API.

**Known limitations / follow-ups:** no nominal (category-specific) or scale effects yet (scale would plausibly map onto `dispformula` later, per @DrJerryTAO's analysis above); DHARMa's automatic path needs a small DHARMa-side fix (its `getObservedResponse()` does `as.numeric(factor) - 1`, a binomial assumption, while simulations come through 1-based — the manual `createDHARMa()` route gives uniform residuals, KS p = 0.86); no emmeans support yet.

Would you be open to a PR? If so, which of the design points above would you like handled differently first?

*Full disclosure: this was developed with substantial help from an AI assistant (Claude Fable 5), with me directing the scope and design (following your Oct 2024 comment). That's part of why I've leaned hard on external validation — every number above is a direct comparison against MASS::polr and ordinal::clm/clmm, and the testthat file encodes those comparisons so they run in CI — and I'm happy to walk through any part of the diff during review.*

**Comment by bbolker** [comment]
Let me think about this a bit. This all seems pretty reasonable, I think I'd be open to a PR.

1. This is OK (it was indeed my choice), but I'd want to try to make sure that violating the previous implicit assumption that `length(psi) < 2` doesn't do anything to any downstream packages (`emmeans`, stuff in `ezstats/parameters`, etc.).
2. This is interesting and a can of worms. I wonder about a "softmax" approach as in Koslik et al 2025?
3. In general I'm nervous about internal `map` settings &mdash; we have to be careful to distinguish internal and user-specified `map` settings in various downstream/output contexts &mdash; but this seems reasonable. Not really worried about the cosmetics, nice to fix if it can be done without lots of code complexity
4. This seems weird (how does it makes sense to take the mean of what are essentially categorical values?) but I don't have a better idea at the moment.


5 and 6 seem fine.


Koslik, Jan-Ole, Fanny Dupont, Marie Auger-Méthé, Marianne Marcoux, Nigel Hussey, and Nancy Heckman. 2025. “Flexible Unimodal Density Estimation in Hidden Markov Models.” arXiv:2511.17071. Preprint, arXiv, November 25. https://doi.org/10.48550/arXiv.2511.17071.

**Comment by strengejacke** [comment]
> 1. This is OK (it was indeed my choice), but I'd want to try to make sure that violating the previous implicit assumption that `length(psi) < 2` doesn't do anything to any downstream packages (`emmeans`, stuff in `ezstats/parameters`, etc.).

We have some special handling for ordinal models (e.g. https://github.com/easystats/insight/blob/66e73e6bb53d56e01ce24a331b09d4cb4c0a5aa3/R/get_parameters.R#L719), but I think it should be rather easy to implement support for glmmTMB ordinal models, since we have experience with all the other cumulative/ordinal/multinomial model classes, and brms

**Comment by jmgirard** [comment]
Awesome! I'll aim to have a PR ready by the end of this weekend.
